### ISSUE TITLE
ttyFIQ0 console

### ISSUE THREAD

**Issue by micsuka** [aloksinha2001/Linux3188#13]
I have a mk908 device (I'm not sure if it's original, but whatever) and I'm trying to use it as a linux server (I'm running Archlinux on it).

I'm trying to use the ttyFIQ0 device as the serial console device (console=ttyFIQ0 kernel parameter). (I've made the hardware hack of the rx/tx pins).

It works, it boots, but when the system (systemd) reaches a point in the boot process, it crashes. (I guess it uses it in a multi-threaded environment).

So, I have a kernel callstack here:
[    2.764919]  ov2659 write reg(0x103, val:0x1) failed, try to write again!
[   24.665130] 1ee0: 0000fe00 00000000 00000000 00000000 00000000 00000000 00001000 00000000
[   24.673314] 1f00: 00000000 00000001 ef0118c8 ef0118c0 00000001 ef0118c0 f007c430 00000000
[   24.681501] 1f20: 00000000 00000000 ffffffff ef0118c0 c082d000 f007c430 efc02770 ef575000
[   24.689687] 1f40: 00000000 00000008 ef0118c8 c090ae70 00000000 00000000 00000005 ef0118c0
[   24.697871] 1f60: 00000000 c240f460 ef0118c0 c0833484 c26a0000 00000000 00000000 c0907438
[   24.706059] 1f80: 00000003 c240f460 c240f4a0 c0907504 00000000 bea8bf5f bea89cfc 00000106
[   24.714245] 1fa0: 00000006 c0833300 bea8bf5f bea89cfc 00000003 0000540e 00000001 00000000
[   24.722428] 1fc0: bea8bf5f bea89cfc 00000106 00000006 00000000 00000002 400e04d0 00000000
[   24.730610] 1fe0: 00000000 bea8969c 0000b04c 402a9bec 400b0010 00000003 adb5b5f2 c3cbefbf
[   24.738807] [<c084073c>](fiq_tty_close+0x4/0x1c) from [<c0a45280>](tty_release+0x118/0x4f8)
[   24.747352] [<c0a45280>](tty_release+0x118/0x4f8) from [<c090ae70>](fput+0xc0/0x1fc)
[   24.755281] [<c090ae70>](fput+0xc0/0x1fc) from [<c0907438>](filp_close+0x58/0x80)
[   24.762949] [<c0907438>](filp_close+0x58/0x80) from [<c0907504>](sys_close+0xa4/0xd8)
[   24.770966] [<c0907504>](sys_close+0xa4/0xd8) from [<c0833300>](ret_fast_syscall+0x0/0x30)
[   24.779410] Code: 0583057c e3a00000 e12fff1e e5903128 (e5932580) 
[   24.785586] ---[ end trace 9e102f68ee9fa529 ]---
[   24.790615] ------------[ cut here ]------------
[   24.795294] WARNING: at drivers/tty/tty_mutex.c:28 tty_lock+0x54/0x68()
[   24.802058] [<c0838e64>](unwind_backtrace+0x0/0x108) from [<c086d164>](warn_slowpath_common+0x54/0x6c)
[   24.811633] [<c086d164>](warn_slowpath_common+0x54/0x6c) from [<c086d198>](warn_slowpath_null+0x1c/0x24)
[   24.821428] [<c086d198>](warn_slowpath_null+0x1c/0x24) from [<c0d5e13c>](tty_lock+0x54/0x68)
[   24.830179] [<c0d5e13c>](tty_lock+0x54/0x68) from [<c0a440a4>](__tty_hangup+0x68/0x3c8)
[   24.838616] [<c0a440a4>](__tty_hangup+0x68/0x3c8) from [<c0a45c10>](disassociate_ctty+0x7c/0x1f8)
[   24.847822] [<c0a45c10>](disassociate_ctty+0x7c/0x1f8) from [<c0871460>](do_exit+0x538/0x788)
[   24.856587] [<c0871460>](do_exit+0x538/0x788) from [<c0836dd4>](die+0x268/0x31c)
[   24.864252] [<c0836dd4>](die+0x268/0x31c) from [<c0d549b8>](__do_kernel_fault.part.3+0x60/0x80)
[   24.873246] [<c0d549b8>](__do_kernel_fault.part.3+0x60/0x80) from [<c083a730>](do_page_fault+0x1d4/0x1ec)
[   24.883154] [<c083a730>](do_page_fault+0x1d4/0x1ec) from [<c082d530>](do_DataAbort+0x38/0xa4)
[   24.891988] [<c082d530>](do_DataAbort+0x38/0xa4) from [<c0832dac>](__dabt_svc+0x4c/0x60)
[   24.900344] Exception stack(0xc26a1e78 to 0xc26a1ec0)
[   24.905423] 1e60:                                                       c2610800 ef0118c0
[   24.913629] 1e80: ef573d80 00000000 c2610800 00000000 ef0118c0 c082d000 00000000 c2610800
[   24.921849] 1ea0: f007c430 ef0118c8 00005401 c26a1ec0 c0a45280 c084073c a00b0013 ffffffff
[   24.930070] [<c0832dac>](__dabt_svc+0x4c/0x60) from [<c084073c>](fiq_tty_close+0x4/0x1c)
[   24.938380] [<c084073c>](fiq_tty_close+0x4/0x1c) from [<c0a45280>](tty_release+0x118/0x4f8)
[   24.946923] [<c0a45280>](tty_release+0x118/0x4f8) from [<c090ae70>](fput+0xc0/0x1fc)
[   24.954869] [<c090ae70>](fput+0xc0/0x1fc) from [<c0907438>](filp_close+0x58/0x80)
[   24.962565] [<c0907438>](filp_close+0x58/0x80) from [<c0907504>](sys_close+0xa4/0xd8)
[   24.970603] [<c0907504>](sys_close+0xa4/0xd8) from [<c0833300>](ret_fast_syscall+0x0/0x30)
[   24.979072] ---[ end trace 9e102f68ee9fa52a ]---

I'm almost sure that the fiq_debugger driver can't be used in a multi-core environment and this crash is a race condition problem.
So actually this issue is a question for you guys: do you know anything more-in-depth about using the ttyFIQ0 device as a serial console?

Or should I use another serial device for serial console? I don't need the fiq debugging itself.

thanks

**Comment by micsuka** [comment]
sorry for pasting the trace in my previous post. Here is it: http://pastebin.com/zfiifKs3

**Comment by micsuka** [comment]
Sorry for the delay - I haven't had time :)

So, I removed the camera driver - I think it has no relation to this problem.

The full boot log is here: http://pastebin.com/EP13GUBr
The kernel config is here: http://pastebin.com/EaC78HfB

What I saw: the fiq driver (arm/common/fiq_debugger.c) has spinlocks in the 3.4 kernel but it doesn't in the 3.0.
(spin_lock_irq(&state->console_lock);..)

So, the 3.0 implementation is probably is not a multiprocessor-safe implementation.
